stride/physics/iso_elastic/devito.py

```diff
@@ -0,0 +1,398 @@
+
+            rec_term = rec_tau.interpolate(expr=tau[0, 0] + tau[1, 1])
+            # rec_term += rec_v1.interpolate(expr=v[1])  # Placeholder for vel_x receiver
+            # rec_term += rec_v2.interpolate(expr=v[0])  # Placeholder for vel_y receiver
+
+            # Set up parameters as functions
+            lam_fun = self.dev_grid.function('lam_fun')
+            mu_fun = self.dev_grid.function('mu_fun')
+            byn_fun = self.dev_grid.function('byn_fun')
+
+            # Compile the operator
+            # velocity (first derivative vel w.r.t. time, first order euler method), s: time_spacing
+            u_v = devito.Eq(vel.forward,
+                            damp*(vel + s * byn_fun * devito.div(tau)),
+                            grid=self.dev_grid,
+                            coefficients=None)
+
+            # stress (first derivative tau w.r.t. time, first order euler method)
+            u_tau = devito.Eq(tau.forward,
+                              damp *
+                              (tau
+                               + s * (
+                                       lam_fun * devito.diag(devito.div(vel.forward))
+                                       + mu_fun * (devito.grad(vel.forward) + devito.grad(vel.forward).T)
+                                      )))
+
+            self.state_operator.set_operator([u_v] + [u_tau] + src_xx + src_zz + rec_term,
+                                             **kwargs)
+            self.state_operator.compile()
+
+        else:
+            # If the source/receiver size has changed, then create new functions for them
+            if num_sources != self.dev_grid.vars.src.npoint:
+                self.dev_grid.sparse_time_function('src', num=num_sources, cached=False)
+
+            if num_receivers != self.dev_grid.vars.rec_tau.npoint:
+                self.dev_grid.sparse_time_function('rec', num=num_receivers, cached=False)
+
+        # Clear all buffers
+        self.dev_grid.vars.src.data_with_halo.fill(0.)
+        self.dev_grid.vars.rec_tau.data_with_halo.fill(0.)
+        self.dev_grid.vars.vel[0].data_with_halo.fill(0.)
+        self.dev_grid.vars.vel[1].data_with_halo.fill(0.)
+        self.dev_grid.vars.tau[0, 0].data_with_halo.fill(0.)
+        self.dev_grid.vars.tau[0, 1].data_with_halo.fill(0.)
+        self.dev_grid.vars.tau[1, 0].data_with_halo.fill(0.)
+        self.dev_grid.vars.tau[1, 1].data_with_halo.fill(0.)
+        self.boundary.clear()
+
+        # Set medium parameters
+        vp_with_halo = self.dev_grid.with_halo(vp.extended_data)
+        vs_with_halo = self.dev_grid.with_halo(vs.extended_data)
+        rho_with_halo = self.dev_grid.with_halo(rho.extended_data)
+
+        lam_with_halo = rho_with_halo * (vp_with_halo ** 2 - 2. * vs_with_halo ** 2)
+        mu_with_halo = self.dev_grid.with_halo(rho_with_halo * vs_with_halo ** 2)
+        byn_with_halo = self.dev_grid.with_halo(1 / rho_with_halo)
```

Comment:
mu and byn do not need the extra "with_halo", rho and vs are already the correct size

Reply:
Correction made